### PR TITLE
Compress notebook output streams before rendering

### LINKED ISSUE #11031: VS Code crashes when running a cell that produces a lot of output
I am trying to print model summary in tensorflow and I think the model is large and it's crashing my notebook. The model is ResNet101.

image

The whole computer comes to a halt, memory usage goes up to 99% and VS Code crashes. I have 16 GB of ram, so I didn't think printing something large would actually eat all my ram. Also, because the kernel crashes, all the variables are lost like history = model.fit() which I need to fine-tune the model afterwards. Moreover, I need to print base_model summary in order to choose from which layer to fine-tune from.

Is there a way to print the summary in another way and save the entire notebook with the variables, so I can continue working? I have checkpoints for model weights, but I need to keep track of past epochs through history to resume training afterwards.

I will try using`.py` files, but I want to know if there is a way to solve this problem for jupyter.

### PR DESCRIPTION
**Pull request by DonJayamanne**
Fixes https://github.com/microsoft/vscode-jupyter/issues/11031
Corresponding PR in Jupyter is here https://github.com/microsoft/vscode-jupyter/pull/11350

### CODE REVIEW

File: extensions/ipynb/src/serializers.ts [hunk 1]
```diff
@@ -270,21 +271,17 @@ type JupyterOutput =
...
-			// If the last item in `outputs` is not empty and the first item in `lines` is not empty, then concate them.
-			// As they are part of the same line.
-			if (outputs.length && lines.length && lines[0].length > 0) {
-				outputs[outputs.length - 1] = `${outputs[outputs.length - 1]}${lines.shift()!}`;
-			}
-			for (const line of lines) {
-				outputs.push(line);
-			}
-		});
+	const compressedStream = output.items.length ? compressOutputItemStreams(output.items[0].mime, output.items).toString() : '';
```
Comment (rebornix) [inline extensions/ipynb/src/serializers.ts:274]: It seems we miss a `.filter((opit) => opit.mime === CellOutputMimeTypes.stderr || opit.mime === CellOutputMimeTypes.stdout)` (which we did before)

File: src/vs/workbench/contrib/notebook/browser/view/renderers/stdOutErrorPreProcessor.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		if (output.mime === mimeType) {
+			if ((buffers.length === 0 || startAppending)) {
+				buffers.push(output.data.buffer);
+				startAppending = true;
+			}
+		} else if (startAppending) {
+			startAppending = false;
+		}
+	}
+	compressStreamBuffer(buffers);
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/view/renderers/stdOutErrorPreProcessor.ts:34]: We are doing `compressStreamBuffer` in serializer, so not sure if we need to do this here. The concern is `stream.subarray(0, MOVE_CURSOR_1_LINE_COMMAND.length).toString()` in a relatively large progress output might be slower than just copying the buffer and sending over to the webview.

**Review by rebornix — COMMENTED** [review]
LGTM in overall, changes are made in right places and look great. Only two questions before we have it merged.

File: src/vs/workbench/contrib/notebook/browser/view/renderers/stdOutErrorPreProcessor.ts (hunk 2 above)
Comment (rebornix) [reply src/vs/workbench/contrib/notebook/browser/view/renderers/stdOutErrorPreProcessor.ts:34]: A manual test of 1000 lines of stdout output (1000 `appendOutputItem`) is probably sufficient

File: extensions/ipynb/src/serializers.ts (hunk 1 above)
Comment (DonJayamanne) [reply extensions/ipynb/src/serializers.ts:274]: I removed that as I don't see how we can have an output with different mime types when dealing with stdout/stderr.
Ideally they should be different toutputs, it cannot be mixed in the same output.

Comment (DonJayamanne) [reply extensions/ipynb/src/serializers.ts:274]: I.e. it wasn't necessary

File: src/vs/workbench/contrib/notebook/browser/view/renderers/stdOutErrorPreProcessor.ts (hunk 2 above)
Comment (DonJayamanne) [reply src/vs/workbench/contrib/notebook/browser/view/renderers/stdOutErrorPreProcessor.ts:34]: >  to the webview (cost of toString() + GC of the short lived string).

Not following you here, there's no `toString` taking place here, we're sending the raw bytes

File: src/vs/workbench/contrib/notebook/browser/view/renderers/stdOutErrorPreProcessor.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	streams.forEach((stream, index) => {
+		if (index === 0 || stream.length < MOVE_CURSOR_1_LINE_COMMAND.length) {
+			return;
+		}
+
+		const previousStream = streams[index - 1];
+
+		// Remove the previous line if required.
+		const command = stream.subarray(0, MOVE_CURSOR_1_LINE_COMMAND.length);
+		if (command[0] === MOVE_CURSOR_1_LINE_COMMAND_BYTES[0] && command[1] === MOVE_CURSOR_1_LINE_COMMAND_BYTES[0] && command[2] === MOVE_CURSOR_1_LINE_COMMAND_BYTES[0]) {
```
Comment (rebornix) [inline src/vs/workbench/contrib/notebook/browser/view/renderers/stdOutErrorPreProcessor.ts:50]: 👏

File: extensions/ipynb/src/streamCompressor.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+function compressStreamBuffer(streams: Uint8Array[]) {
+	streams.forEach((stream, index) => {
+		if (index === 0 || stream.length < MOVE_CURSOR_1_LINE_COMMAND.length) {
+			return;
+		}
+
+		const previousStream = streams[index - 1];
+
+		// Remove the previous line if required.
+		if (stream.subarray(0, MOVE_CURSOR_1_LINE_COMMAND.length).toString() === MOVE_CURSOR_1_LINE_COMMAND) {
```
Comment (rebornix) [inline extensions/ipynb/src/streamCompressor.ts:51]: maybe we can compare bytes here too.

**Review by rebornix — DISMISSED** [review]

File: extensions/ipynb/src/streamCompressor.ts (hunk 4 above)
Comment (DonJayamanne) [reply extensions/ipynb/src/streamCompressor.ts]: Yes, done

File: extensions/ipynb/src/serializers.ts [hunk 5]
```diff
@@ -4,9 +4,11 @@
  *--------------------------------------------------------------------------------------------*/
 
 import type * as nbformat from '@jupyterlab/nbformat';
+import { TextDecoder } from 'util';
```
Comment (rebornix) [inline extensions/ipynb/src/serializers.ts:7]: I think we don't need this import, not sure if having it explicit would break the browser build.

**Review by rebornix — APPROVED** [review]

**Pull request merged**